### PR TITLE
feat(service-offering): show all service offerings even those which d…

### PR DESCRIPTION
**Pull request by HeyRoach**
…on't fir account resource limitation

### CODE REVIEW

File: src/app/service-offering/custom-service-offering/custom-service-offering.component.ts [hunk 1]
```diff
@@ -1,24 +1,46 @@
...
     @Inject(MAT_DIALOG_DATA) data,
     public dialogRef: MatDialogRef<CustomServiceOfferingComponent>,
   ) {
-    const { offering } = data;
-    this.offering = offering;
+    this.offering = data.offering;
+    this.account = data.account;
+  }
+
+  public ngOnInit() {
```
Comment (tamazlykar) [inline src/app/service-offering/custom-service-offering/custom-service-offering.component.ts]: Why you use a component hook to do this instead if constructor?

File: src/app/service-offering/custom-service-offering/custom-service-offering.component.ts [hunk 2]
```diff
@@ -35,11 +57,10 @@ export class CustomServiceOfferingComponent {
...
 
   private createForm() {
-    // input text=number provide all other validation for current restrictions
     this.hardwareForm = new FormGroup({
-      cpuNumber: new FormControl(this.offering.cpunumber, Validators.required),
-      cpuSpeed: new FormControl(this.offering.cpuspeed, Validators.required),
-      memory: new FormControl(this.offering.memory, Validators.required),
+      cpuNumber: new FormControl(this.offering.cpunumber, [Validators.required, LimitValidator(this.maxCpu)]),
+      cpuSpeed: new FormControl(this.offering.cpuspeed, [Validators.required]),
+      memory: new FormControl(this.offering.memory, [Validators.required, LimitValidator(this.maxMemory)]),
```
Comment (tamazlykar) [inline src/app/service-offering/custom-service-offering/custom-service-offering.component.ts]: The code does not work without this check? 
The comment `// input text=number provide all other validation for current restrictions` describe that all validation provided by input itself. I think it should be placed back. You add only validation for max value, but input provide both max and min. See #1082

File: src/app/service-offering/custom-service-offering/custom-service-offering.component.ts [hunk 3]
```diff
@@ -1,24 +1,46 @@
...
+import { FormControl, FormGroup, ValidatorFn, Validators } from '@angular/forms'
 import { MAT_DIALOG_DATA, MatDialogRef } from '@angular/material';
 
 import { ComputeOfferingViewModel } from '../../vm/view-models';
+import { Account } from '../../shared/models';
+
+function LimitValidator(cpuNumberLimit: number): ValidatorFn {
+  return function (control: FormControl) {
+    return control.value > cpuNumberLimit
+      ? { cpuLimitExceeded: true }
```
Comment (tamazlykar) [inline src/app/service-offering/custom-service-offering/custom-service-offering.component.ts]: This is the same as built-in validator  `Validators.max(max: number)`

File: src/app/service-offering/service-offering-dialog/service-offering-dialog.component.html [hunk 4]
```diff
@@ -40,7 +42,7 @@ <h3 class="mat-dialog-title">
       <button
         mat-button
         color="primary"
-        [disabled]="isSubmitButtonDisabled()"
+        [disabled]="isSubmitButtonDisabled() || resourcesLimitExceeded"
```
Comment (tamazlykar) [inline src/app/service-offering/service-offering-dialog/service-offering-dialog.component.html]: Move to correspond function

File: src/app/service-offering/service-offering-dialog/service-offering-dialog.component.html [hunk 5]
```diff
@@ -14,9 +14,11 @@ <h3 class="mat-dialog-title">
       [classes]="classes"
       [selectedClasses]="selectedClasses"
       [query]="query"
+      [account]="account"
       [offeringList]="serviceOfferings"
       [selectedOffering]="serviceOffering"
       [showFields]="showFields"
+      [resourcesLimitExceeded]="resourcesLimitExceeded"
```
Comment (tamazlykar) [inline src/app/service-offering/service-offering-dialog/service-offering-dialog.component.html]: Why do you decide to show this error in service-offering-list but not here?

File: src/app/service-offering/service-offering-dialog/service-offering-dialog.component.ts [hunk 6]
```diff
@@ -99,4 +103,18 @@ export class ServiceOfferingDialogComponent implements OnInit, OnChanges {
     return isDifferentOfferingId || isSameCustomOfferingWithDifferentParams;
   }
 
+  private checkLimits(offering: ComputeOfferingViewModel) {
+    let сpusExceeded;
+    let memoryExceeded;
+
+    if (offering.iscustomized) {
+      сpusExceeded = this.account.cpuavailable < this.serviceOffering['customOfferingRestrictions'].cpunumber.min;
+      memoryExceeded = this.account.memoryavailable < this.serviceOffering['customOfferingRestrictions'].memory.min;
```
Comment (tamazlykar) [inline src/app/service-offering/service-offering-dialog/service-offering-dialog.component.ts]: Why you use square brackets to get `customOfferingRestrictions` field?

File: src/app/vm/selectors/service-offering.selectors.ts [hunk 7]
```diff
@@ -61,21 +61,7 @@ const getAvailableByResourcesSync = (
...
-      enoughCpus = resourceUsage.available.cpus >= offering.customOfferingRestrictions.cpunumber.min;
-      enoughMemory = resourceUsage.available.memory >= offering.customOfferingRestrictions.memory.min;
-    } else {
-      enoughCpus = resourceUsage.available.cpus >= offering.cpunumber;
-      enoughMemory = resourceUsage.available.memory >= offering.memory;
-    }
-
-    return enoughCpus && enoughMemory;
-  });
+  return availableInZone;
```
Comment (tamazlykar) [inline src/app/vm/selectors/service-offering.selectors.ts]: Why you do keep this function which only uses another function?

File: src/i18n/ru.json [hunk 8]
```diff
@@ -118,6 +118,9 @@
     },
     "SNAPSHOT_POLICIES": {
       "HOURLY_TURN_OFF": "Почасовое расписание отключено"
+    },
+    "COMPUTE_OFFERING": {
+      "RESOURCE_LIMIT_EXCEEDED": "Вычислительное предложение не может быть выбрано, потому что больше, чем доступные для аккаунта ресурсы."
```
Comment (tamazlykar) [inline src/i18n/ru.json]: Can we improve this translation? To discuss.

File: src/app/service-offering/custom-service-offering/custom-service-offering.component.ts (hunk 1 above)
Comment (HeyRoach) [reply src/app/service-offering/custom-service-offering/custom-service-offering.component.ts]: Mostly we use ngOnInit for all the initialization/declaration and avoid stuff to work in the constructor. The constructor should only be used to initialize class members but shouldn't do actual "work".

File: src/app/service-offering/custom-service-offering/custom-service-offering.component.ts (hunk 2 above)
Comment (HeyRoach) [reply src/app/service-offering/custom-service-offering/custom-service-offering.component.ts]: To discuss

File: src/app/service-offering/custom-service-offering/custom-service-offering.component.scss [hunk 9]
```diff
@@ -9,3 +9,8 @@ h5 {
     margin-top: 0;
   }
 }
+
+.error-message {
+  color: red;
```
Comment (tamazlykar) [inline src/app/service-offering/custom-service-offering/custom-service-offering.component.scss]: Use color from `mat-error`

File: src/app/service-offering/custom-service-offering/custom-service-offering.component.ts [hunk 10]
```diff
@@ -1,25 +1,29 @@
-import { Component, Inject } from '@angular/core';
-import { FormControl, FormGroup, Validators } from '@angular/forms'
+import { Component, Inject, OnInit } from '@angular/core';
+import { FormControl, FormGroup, ValidatorFn, Validators } from '@angular/forms'
```
Comment (tamazlykar) [inline src/app/service-offering/custom-service-offering/custom-service-offering.component.ts]: `ValidatorFn` unused

File: src/app/service-offering/service-offering-dialog/service-offering-dialog.component.scss [hunk 11]
```diff
@@ -20,3 +20,8 @@
   flex-direction: column;
   text-align: right;
 }
+
+.error-message {
+  color: red;
```
Comment (tamazlykar) [inline src/app/service-offering/service-offering-dialog/service-offering-dialog.component.scss]: Color

File: src/app/service-offering/service-offering-dialog/service-offering-dialog.component.ts [hunk 12]
```diff
@@ -81,7 +83,19 @@ export class ServiceOfferingDialogComponent implements OnInit, OnChanges {
...
-      || isSelectedOfferingDifferentFromCurrent;
+      || isSelectedOfferingDifferentFromCurrent
+      || isNotEnoughResourcesForCurrentOffering;
+  }
+
+  public isSelectedOfferingViewMode(): boolean {
+    if (this.serviceOffering.iscustomized && this.viewMode === ServiceOfferingType.custom) {
+      return true;
+    }
+
```
Comment (tamazlykar) [inline src/app/service-offering/service-offering-dialog/service-offering-dialog.component.ts]: Blank line

File: src/app/vm/selectors/view-models/compute-offering-view-model.selector.spec.ts [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+import { account } from '../../../../testutils/data/accounts';
+import { nonCustomizableProperties } from '../../../core/config/default-configuration';
+import { ComputeOfferingViewModel } from '../../view-models';
+import { Account } from '../../../shared/models';
+import {
+  CustomComputeOfferingParameters
+} from '../../../shared/models/config/custom-compute-offering-parameters.interface';
+
+describe('ComputeOfferingViewModelSelector', () => {
+  describe('isAvailableByResources', () => {
```
Comment (tamazlykar) [inline src/app/vm/selectors/view-models/compute-offering-view-model.selector.spec.ts]: Do we need this describes?

File: src/app/vm/selectors/view-models/compute-offering-view-model.selector.spec.ts [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+  });
+
+
+  // it('isAvailableByResources in custom compute offering params which satisfy resource should be true', () => {
+  //   const customComputeOfferingParameters = [
+  //     {
+  //       'offeringId': '36de12ed-17f1-441f-903f-ab274832c318',
+  //       'cpunumber': {
+  //         'min': 2,
+  //         'max': 8,
```
Comment (tamazlykar) [inline src/app/vm/selectors/view-models/compute-offering-view-model.selector.spec.ts]: Commented code

File: src/app/vm/vm-creation/components/service-offering-selector/service-offering-selector.component.scss [hunk 15]
```diff
@@ -18,3 +18,8 @@
   text-align: left;
   padding-top: 18px;
 }
+
+.error-message {
+  color: red;
```
Comment (tamazlykar) [inline src/app/vm/vm-creation/components/service-offering-selector/service-offering-selector.component.scss]: color

File: src/app/vm/vm-creation/containers/vm-creation.container.ts [hunk 16]
```diff
@@ -2,7 +2,7 @@ import { Component, OnInit } from '@angular/core';
 import { MatDialogRef } from '@angular/material';
 import { select, Store } from '@ngrx/store';
 import { combineLatest, Observable } from 'rxjs';
-import { first, map } from 'rxjs/operators';
+import { first, filter, map } from 'rxjs/operators';
```
Comment (tamazlykar) [inline src/app/vm/vm-creation/containers/vm-creation.container.ts]: `filter` unused

File: src/app/vm/selectors/view-models/compute-offering-view-model.selector.ts [hunk 17]
```diff
@@ -237,5 +226,70 @@ export const getComputeOfferingViewModel = createSelector(
...
+  configSelectors.get('defaultCustomComputeOfferingRestrictions'),
+  configSelectors.get('customComputeOfferingHardwareValues'),
+  UserTagsSelectors.getServiceOfferingParamTags,
+  fromVms.getSelectedVM,
+  (account,
+   offerings,
+   customComputeOfferingParameters,
+   defaultRestrictions,
+   defaultHardwareValues,
+   tags, vm): ComputeOfferingViewModel[] => {
```
Comment (tamazlykar) [inline src/app/vm/selectors/view-models/compute-offering-view-model.selector.ts]: Put each parameter on new line, or all parameters in one line

File: src/app/vm/selectors/view-models/compute-offering-view-model.selector.ts [hunk 18]
```diff
@@ -237,5 +226,70 @@ export const getComputeOfferingViewModel = createSelector(
...
+
+    const cpuNumber = account && account.cpuavailable === 'Unlimited'
+      ? account.cpuavailable
+      : Number(account.cpuavailable) + cpuNumberUsed;
+    const memory = account && account.memoryavailable === 'Unlimited'
+      ? account.memoryavailable
+      : Number(account.memoryavailable) + memoryUsed;
+
+    const availableResources: Resources = {
+      cpuNumber: cpuNumber || cpuNumberUsed,
```
Comment (tamazlykar) [inline src/app/vm/selectors/view-models/compute-offering-view-model.selector.ts]: Add comment about fallback value

File: src/app/vm/selectors/view-models/compute-offering-view-model.selector.ts [hunk 19]
```diff
@@ -237,5 +226,70 @@ export const getComputeOfferingViewModel = createSelector(
...
+  configSelectors.get('customComputeOfferingHardwareValues'),
+  UserTagsSelectors.getServiceOfferingParamTags,
+  (account,
+   offerings,
+   customComputeOfferingParameters,
+   defaultRestrictions,
+   defaultHardwareValues,
+   tags): ComputeOfferingViewModel[] => {
+    const availableResources: Resources = {
+      cpuNumber: account && account.cpuavailable || '0',
```
Comment (tamazlykar) [inline src/app/vm/selectors/view-models/compute-offering-view-model.selector.ts]: Add comment about fallback value

**Pull request merged**